[Kitsune] Added vecscale to test suite benchmarks

jsarrao opened the pull request:
Vecscale is a simple benchmark that generates a random vector and subsequently scales every element by a constant value.

tarunprabhu committed:
[Kitsune] Add end-to-end-tests for Kitsune

Add end-to-end tests for Kitsune. The initial tests included here are from the
kitsune/experiments directory of the Kitsune repo. These have been divided into
three categories:

- Benchmarks
- SingleSource
- MultiSource

The benchmarks are single-source tests consisting of multiple implementations in
a range of languages. These include cuda, hip, Kokkos (with simple parallel_for
constructs), Kokkos with DualView's in addition to the Kitsune-specific forall.
These are intended to be used for both correctness checks and performance
comparisons. The implementations share as much code as possible - mostly to
read command line arguments. As a result the software engineering is somewhat
dreadful, but the goal was to enable easy experimentation by providing
convenient command line arguments to change the problem size across all of the
implementations. Each of the tests have two different problem sizes associated
with it. One is very small and executes in under a second to allow for quick
correctness checks. A larger problem is more suitable for performance
comparisons.

The tests in SingleSource/ are exactly what the name suggests - simple code
contained within a single file. They are intended to be used purely for test for
end-to-end correctness. The tests in kitsune's repository now perform many of
the same tests but in a piecemeal fashion. However, the tests in SingleSource/
are still useful since act as a sanity check over and above the granular tests.

The tests in MultiSource/ are intended to test separate compilation and LTO.
After the benchmarks, these are, arguably, the most useful since they can be
used to check "real" code that is hard to do in the core LLVM tests.

Currently, there is no support for Fortran, although some hooks have been added.
It is unlikely that we can just add Fortran files into the mix and have them
just work.

A simple timing class and another that performs floating point comparison have
been added. These could almost certainly do with improvement - especially the
timing class, but for now, they will do.

Some scripts have been provided that collect compilation and execution
statistics, both from the default report generated by the LLVM test-suite
infrastructure itself, the timings reported by the tests themselves and from
the timeit script that wraps all compilations in the test suite. This produces
a report in JSON format, but that is not the most user-friendly. Over time,
more scripts should be added that can parse that report and generate graphs or
other summaries that focus on specific characteristics.

This is also intended to supersede the code in kitsune/experiments. To that
end, some code has been imported from there that is not actually used yet. This
is intentional.

This commit also adds CPU and GPU reference outputs for the benchmarks. These
are different because the floating point error is different in the two cases.
This seems to be a function of both the frontend and backend optimizations
carried out by the compiler on these platforms. It is not clear that we can
rely on a single "expected" output for some of the more complex code.

This is a squash of commits spanning several months. There didn't seem to be
any reason to keep all those individual commits around since there was a fair
bit of churn to keep up with the significant changes taking place in parallel
in Kitsune's frontend.

tarunprabhu committed:
[Kitsune] Fix bug that caused failures on AArch64

Assigning an integer value to a variable of type char behaves differently on
AArch64 since char's are unsigned on the latter. When assigning -1 to a char
results in the char having the value 255 on AArch64 while it is -1 on X86. The
code relied on the variable having the value -1. Changing the type of the
variable to an int, thereby matching the type of the RHS, fixes the issue.

jsarrao committed:
[Kitsune] Added vecscale to test suite benchmarks

jsarrao committed:
[vecscale] fixed some leftover references to vecadd

jsarrao commented:
I haven't tested any of this, especially because my laptops version of kitsune is not fully functional. I will of course test it if/when I get my new ztoken, but I figured I'd make a pull request in case you wanted to test it out before I get a chance to

Commit:
[Kitsune] Bugfixes to code and CMakeLists.txt

Add synchronize after kernel launch in euler3d.cu

Handle the TEST_TAPIR_TARGETS variable correctly when attempting to iterate
over its contents.

Commit:
[Kitsune] Record compile time statistics even if test fails

Earlier, if the test failed, no statistics were recorded. Here, record the
compile-time statistics even if the test fails at runtime. It is not clear when
this is useful, but it might be.

Commit:
[Kitsune] Run shfmt on run-tests and use lscpu

lscpu may give more consistent information about the CPU than simply looking up
/proc/cpuinfo. At least on one AArch64 machine, /proc/cpuinfo did not contain
any model or vendor information at all while lscpu did provide this.

Run shfmt to indent with spaces instead of tabs.

jsarrao committed:
[Kitsune] Add utility to generate plots from a test suite report

kit-plot is a utility that generates plots from a Kitsune test suite report.
There are 3 main - and several, less significant - options that a user can
provide: benchmarks, metrics, and targets. Benchmarks are the benchmarks the
test suite runs (vecadd, srad, euler3d, etc.), metrics are the types of data
collected by the test suite (compile time, runtime, etc.), and targets refers to
the combination of tapir targets and "external" compilers the test suite tests.
For each metric, kit-plot will generate a grouped bar chart showing a comparison
of the different targets across all benchmarks.

tarunprabhu commented on Kitsune/Benchmarks/vecscale/vecscale.cu:
```diff
@@ -0,0 +1,41 @@
+__global__ void vecscale(const ElementType *a, const ElementType *b, size_t n) {
+  size_t i = blockDim.x * blockIdx.x + threadIdx.x;
+  if (i < n)
+    b[i] = a[i] * 65;
```
Please make `65` a constant in `vecscale.inc` and reference that constant here instead of hard-coding the value everywhere.

tarunprabhu reviewed (changes requested):
Thanks. This looks reasonable at first glance.

Did you include `fpcmp.h` and `timing.h` alongside the benchmark code? They should be symlinks to the headers that are in a common location. That is how it is done in the other benchmarks.

jsarrao committed:
[vecscale] Small changes and testing
  - moved 65 to constant value in vecscale.inc
  - modified Kitsune/utils/CMakeLists to copy kit-plot into test suite build directory
  - modified kit-plot to accept new benchmark
  - fixed file types for fpcmp.h and timing.h
  - verified that vecscale benchmark builds and runs on both cuda and hip gpus
